core/src/main/java/com/linecorp/armeria/common/metric/PrometheusMetricsExporter.java

```diff
@@ -0,0 +1,383 @@
+     * @param prefix the prefix of the exported {@link Metric}s
+     */
+    public PrometheusMetricsExporter(CollectorRegistry registry, String... prefix) {
+        this.registry = requireNonNull(registry, "registry");
+
+        requireNonNull(prefix, "prefix");
+        final StringBuilder buf = new StringBuilder();
+        for (String p : prefix) {
+            requireNonNull(p, "prefix contains null.");
+            checkArgument(!p.isEmpty(), "prefix contains an empty string.");
+            buf.append(p).append('_');
+        }
+        this.prefix = Collector.sanitizeMetricName(buf.toString());
+    }
+
+    @Override
+    protected void exportGauge(Gauge gauge) {
+        register(new PrometheusGauge(gauge));
+    }
+
+    @Override
+    protected void exportDoubleGauge(DoubleGauge gauge) {
+        register(new PrometheusDoubleGauge(gauge));
+    }
+
+    @Override
+    protected void exportHistogram(Histogram histogram) {
+        register(new PrometheusHistogram(histogram));
+    }
+
+    private void register(PrometheusMetric<?> child) {
+        synchronized (collectors) {
+            CollectorImpl c = collectors.get(child.name);
+            if (c == null) {
+                c = new CollectorImpl(child);
+                registry.register(c);
+                collectors.put(child.name, c);
+            } else {
+                c.add(child);
+            }
+        }
+    }
+
+    @VisibleForTesting
+    String prometheusName(MetricKey key, MetricUnit unit) {
+        final StringBuilder buf = new StringBuilder(prefix.length() + 64).append(prefix);
+
+        for (String p : key.nameParts()) {
+            if (ALPHANUM_ONLY_PATTERN.matcher(p).matches()) {
+                buf.append(CaseFormat.LOWER_CAMEL.to(CaseFormat.LOWER_UNDERSCORE, p));
+            } else {
+                buf.append(p);
+            }
+            buf.append('_');
+        }
+
+        if (buf.length() == 0) {
+            throw new IllegalArgumentException("can't specify an empty name when prefix is empty");
+        }
+
+        switch (unit) {
+            case BYTES:
+            case BYTES_CUMULATIVE:
+                buf.append("bytes");
+                break;
+            case NANOSECONDS:
+            case NANOSECONDS_CUMULATIVE:
+                buf.append("seconds");
+                break;
+            default:
+                // Remove the trailing underscore ('_')
+                buf.setLength(buf.length() - 1);
+        }
+
+        if (unit.isCumulative()) {
+            buf.append("_total");
+        }
+
+        return Collector.sanitizeMetricName(buf.toString());
+    }
+
+    static double convertValue(double value, MetricUnit unit) {
+        if (unit == MetricUnit.NANOSECONDS || unit == MetricUnit.NANOSECONDS_CUMULATIVE) {
+            return value / ONE_SECOND_IN_NANOS;
+        } else {
+            return value;
+        }
+    }
+
+    static String sanitizeLabelName(String name) {
```

Comment:
The sanitization isn't that obvious - add a comment? Replaces any non-alphabet letters at the beginning with _ and any non-alphanumeric letters elsewhere with _.